src/imitation/algorithms/sqil.py

```diff
@@ -0,0 +1,215 @@
+    """Replay buffer used in off-policy algorithms like SAC/TD3, modified for SQIL."""
+
+    def __init__(
+        self,
+        buffer_size: int,
+        observation_space: spaces.Space,
+        action_space: spaces.Space,
+        demonstrations: algo_base.AnyTransitions,
+        device: Union[th.device, str] = "auto",
+        n_envs: int = 1,
+        optimize_memory_usage: bool = False,
+    ):
+        """A modification of the SB3 ReplayBuffer.
+
+        This buffer is fundamentally the same as ReplayBuffer,
+        but it includes an expert demonstration internal buffer.
+        When sampling a batch of data, it will be 50/50 expert and collected data.
+
+        Args:
+            buffer_size: Max number of element in the buffer
+            observation_space: Observation space
+            action_space: Action space
+            demonstrations: Expert demonstrations.
+            device: PyTorch device.
+            n_envs: Number of parallel environments. Defaults to 1.
+            optimize_memory_usage: Enable a memory efficient variant
+                of the replay buffer which reduces by almost a factor two
+                the memory used, at a cost of more complexity.
+        """
+        super().__init__(
+            buffer_size,
+            observation_space,
+            action_space,
+            device,
+            n_envs,
+            optimize_memory_usage,
+            handle_timeout_termination=False,
+        )
+
+        self.set_demonstrations(demonstrations)
+
+    def set_demonstrations(
+        self,
+        demonstrations: algo_base.AnyTransitions,
+    ) -> None:
+        """Set the demonstrations to be used in the buffer.
+
+        Args:
+            demonstrations (algo_base.AnyTransitions): Expert demonstrations.
+        """
+        # If demonstrations is a list of trajectories,
+        # flatten it into a list of transitions
+        if not isinstance(demonstrations, types.TransitionsMinimal):
+            (
+                item,
+                demonstrations,
+            ) = util.get_first_iter_element(  # type: ignore[assignment]
+                demonstrations,  # type: ignore[assignment]
+            )
+            if isinstance(item, types.Trajectory):
+                demonstrations = rollout.flatten_trajectories(
+                    demonstrations,  # type: ignore[arg-type]
+                )
+            else:  # item is a TransitionMapping
+                demonstrations = rollout.flatten_transition_mappings(
+                    demonstrations,  # type: ignore[arg-type]
+                )
+
+        assert isinstance(demonstrations, types.Transitions)
+
+        n_samples = len(demonstrations)
+        self.expert_buffer = buffers.ReplayBuffer(
+            n_samples,
+            self.observation_space,
+            self.action_space,
+            handle_timeout_termination=False,
+        )
+
+        for transition in demonstrations:
```

Comment:
Why can't this loop be replaced by a simple `self.expert_buffer.set_demonstrations()`?